### DISCUSSION TITLE
Get changed item from streaming event

### DISCUSSION THREAD

**Discussion by gerardsyd** [ecederstrand/exchangelib#928] [Q&A] [answered]
Hi all,
I've just started using this and managed to get a folder streaming and then running a number of tasks off the back of it. 
Is there a way to get the actual email from the streaming event? At the moment I am using the event as a trigger and then filtering to get the last email to then run the remainder of my logic but I've realised that, at times, the stream sends multiple ModifiedEvent notifications for each email received. 
Snippet of code below:
```
subscription_id = folder.subscribe_to_streaming()
    logging.info('Checking for new messages')
    while True:
        for notification in folder.get_streaming_events(subscription_id, connection_timeout=1):
            for event in notification.events:
                logging.debug(event)
                if isinstance(event, ModifiedEvent):
                    # filter last email from folder
```

**Comment by ecederstrand** [comment, marked as answer]
A `ModifiedEvent` object has an `item_id` attribute that you can use to fetch the message object: `message = folder.get(id= event.item_id.id)`

**Reply by gerardsyd** [reply]
Thanks! I thought I tried that and it asked for more than item id. I'll try again. Many thanks!

**Reply by gerardsyd** [reply]
Hi - so I tried this and it doesn't work. item_id for the Modified Event is None so event.item_id.id throws an error as item_id is a NoneType.

**Reply by ecederstrand** [reply]
Just to rule out XML parsing errors - is the `ItemId` element on the `ModifiedEvent` also missing in the response XML? See https://ecederstrand.github.io/exchangelib/#troubleshooting on how to capture the XML response.

Alternatively, maybe that specific `ModifiedEvent` concerns a folder event, in which case the `item_id` attribute will be empty and `folder_id` will be non-null instead. Just skip those if you don't care about folder events.

**Reply by gerardsyd** [reply]
Oh right, I see. It's picking up folder events since I have it set on a sub-folder and trying to capture new emails that come in that are moved to that folder by Outlook. Was trying to lessen the load on the python app so it didn't get triggered for every email but perhaps that's the better way to do it and just get the app to test for sender. 
Thanks for the quick responses, makes sense.

**Reply by gerardsyd** [reply]
Also am I right in saying that ItemIDs are not unique? For example, I ran a quick script to pull item IDs of all emails in a folder and there were duplicate IDs for different items.

**Reply by ecederstrand** [reply]
Item IDs are unique, at least within an Exchange server: https://docs.microsoft.com/en-us/exchange/client-developer/exchange-web-services/ews-identifiers-in-exchange

If you found duplicate IDs, then it's almost certainly a bug in your code for finding duplicates :-)

**Reply by gerardsyd** [reply]
Hmmm. This is the code I used to get the IDs

```
    filtered_items = folder.all().order_by('-datetime_received')
    df = pd.DataFrame(columns=['ID', 'Date', 'Subject'])0
    for item in filtered_items:
        print(item.id, item.datetime_received, item.subject)
        df = df.append({'ID': item.id, 'Date': item.datetime_received,
                        'Subject': item.subject}, ignore_index=True)
    df.to_csv('test.csv')
```

Actually, you know what, it is a case sensitivity issue. So there is a difference in the ID but it is case sensitive. Let me see if there's a way to solve that in MySQL. Many thanks!
